src/ucs/type/serialize.h

```diff
@@ -0,0 +1,73 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifndef UCS_SERIALIZE_H
+#define UCS_SERIALIZE_H
+
+#include <ucs/sys/compiler_def.h>
+
+/*
+ * Helper macros for serializing/deserializeing custom data
+ */
+
+/**
+ * Advance '_ptr' to the next element, and return a typed pointer to the current
+ * element.
+ *
+ * @param _ptr   Untyped pointer to the current element.
+ * @param _type  Type of the current element.
+ * @param _size  Size of the current element.
+ *
+ * @return Typed pointer to the current element.
+ */
+#define ucs_serialize_next_ptr(_ptr, _type, _size) \
+    ({ \
+        _type *_result = (_type*)(_ptr); \
+        (_ptr)         = UCS_PTR_BYTE_OFFSET(_ptr, _size); \
+        _result; \
+    })
+
+
+/**
+ * Advance '_ptr' to the next element and set a typed pointer to point to the
+ * current element.
+ * .
+ *
+ * @param _ptr        Untyped pointer to the current element.
+ * @param _typed_ptr  Typed pointer, set to point to the current element.
+ */
+#define ucs_serialize_next_typed_ptr(_ptr, _typed_ptr) \
+    { \
+        _typed_ptr = ucs_serialize_next_ptr(_ptr, typeof(*_typed_ptr), \
+                                            sizeof(*_typed_ptr)); \
+    }
+
+
+/**
+ * Set the current element to '_value' and advance to the next element.
+ *
+ * @param _ptr     Untyped pointer to the current element.
+ * @param _type    Type of the current element.
+ * @param _value   Value to set to the current element.
+ */
+#define ucs_serialize_put(_ptr, _type, _value) \
+    { \
+        *ucs_serialize_next_ptr(_ptr, _type, sizeof(_type)) = (_value); \
```

Comment:
`ucs_serialize_next_typed_ptr`?

Reply:
unfortunately, it would be more code, since ucs_serialize_next_typed_ptr expects a typed pointer as an argument, and we don't have such variable in this context